types/index.d.ts

```diff
@@ -938,12 +939,27 @@ export class KafkaJSDeleteGroupsError extends KafkaJSError {
   constructor(e: Error | string, groups?: KafkaJSDeleteGroupsErrorGroups[])
 }
 
+export class KafkaJSDeleteTopicRecordsError extends KafkaJSError {
+  constructor(metadata: KafkaJSDeleteTopicRecordsErrorTopic)
+}
+
 export interface KafkaJSDeleteGroupsErrorGroups {
   groupId: string
   errorCode: number
   error: KafkaJSError
 }
 
+
+export interface KafkaJSDeleteTopicRecordsErrorTopic {
+  topic: string,
+  brokers: KafkaJSDeleteTopicRecordsErrorBroker[]
```

Comment:
Small tweak on your initial suggestion as the requests and responses are at "broker-level". Might be a little less relevant to the end user as they're not exposed to the brokers directly? Though this makes it much more straightforward to map the error to the broker, rather than trying to figure out a way to map the error from the broker to the partition that caused the problem?

Let me know what you think..

Reply:
> Might be a little less relevant to the end user as they're not exposed to the brokers directly?

I agree that it's a bit of a pain in the ass to map, but I really think it's better for the user if we map it to partitions rather than brokers, as the user has sent in a list of partitions and now needs to know which partitions failed. Which broker the requests were made towards is pretty irrelevant.

I guess it's something like this:
```js
const deleteTopicRecordsResults = [
  { partitions: [{ partition: 0, offset: 10 }], error: new Error() },
  { partitions: [{ partition: 1, offset: 10 }, { partition: 2, offset: 10 }], error: new Error() },
]

const resultByPartition = deleteTopicRecordsResults.reduce((result, { partitions, error }) => {
  for (let partition of partitions) {
    result.partitions.push({ ...partition, error })
  }
  return result
}, { topic, partitions: [] })

const error = new KafkaJSDeleteTopicRecordsError({ result })
```

Reply:
Yes, you're right, actually, I realised also if I change how the response is being parsed, I should be able to map the error to the partition precisely (explained in the other comment I just left, on `createErrorFromCode`)